### PR TITLE
Bitmap alloc

### PR DESCRIPTION
**Pull request by djichthys**
A simple bitmap allocator with *CHERI* intrinsics.

### CODE REVIEW

**Comment by ltratt** [comment]
I'm confused by the two commits here -- how does this relate to @jeremysinger's work?

**Comment by djichthys** [comment]
This is the bitmap allocator that @jeremysinger wrote and I made *tiny* modifications to.

**Comment by ltratt** [comment]
OK, but this PR shouldn't have a commit that Jeremy authored then? Basically I'm confused about what changes are new or not. I *think* you might want to rebase this against master, but I'm not sure if that's sufficient or not. Does doing so make things better?

**Comment by djichthys** [comment]
The bitmap allocator is the third memory allocator that @jeremysinger was writing. The *bump-allocator* and the *freelist-allocator* have been merged. 

@jeremysinger wrote the first part of this one i.e. commit 1. I just added the second part to it (commit 2). If acceptable, it will probably be squashed. This PR was raised after cleaning up the branch and rebasing  against [cheri-examples/master](https://github.com/capablevms/cheri-examples/master).  So it should show only the changes with respect to the latest master branch, I hope.

**Comment by ltratt** [comment]
Ah, I think I understand now.

File: example_allocators/bitmap_allocator/README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Simple bitmap allocator implementation, to allocate
```
Comment (ltratt) [inline example_allocators/bitmap_allocator/README.md:1]: This file needs reformatting to consistent line widths (e.g. 80/100 chars).

File: example_allocators/bitmap_allocator/bitmap_alloc.c [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+void init_alloc(int num_chunks, int chunk_size)
+{
+	int i = 0;
+
+	/* we need to increase the num_chunks
+	 * so every bit in bitmap will be used
+	 */
+	int adjusted_num_chunks = (num_chunks % BITS_PER_BYTE == 0)
+								  ? num_chunks
```
Comment (ltratt) [inline example_allocators/bitmap_allocator/bitmap_alloc.c:61]: The large indentation here makes this very difficult to read.

File: example_allocators/bitmap_allocator/bitmap_alloc.c [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	int adjusted_chunk_size =
+		(chunk_size % (sizeof(void *)) == 0)
+			? chunk_size
+			: (chunk_size + (sizeof(void *)) - (chunk_size % (sizeof(void *))));
+
+	/* check this chunk size is small enough
+	 * so we can represent bounds precisely with
+	 * CHERI compressed representation
+	 */
+	// FIXME
```
Comment (ltratt) [inline example_allocators/bitmap_allocator/bitmap_alloc.c]: Should we add a simple check here and then `err` if the check fails?

File: example_allocators/bitmap_allocator/bitmap_alloc.c [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+	buffer_size = adjusted_num_chunks * adjusted_chunk_size;
+	bitmap_size = adjusted_num_chunks / BITS_PER_BYTE;
+
+	/* zero bitmap, since all chunks are free initially */
+	for (i = 0; i < bitmap_size; i++)
+	{
+		bitmap[i] = 0;
+	}
+
+	// FIXME - should we use CHERI API to
```
Comment (ltratt) [inline example_allocators/bitmap_allocator/bitmap_alloc.c]: We can't set *exact* bounds in all cases but we can set a safe approximate upper limit (i.e. we over-approximate in the worst case), right?

File: example_allocators/bitmap_allocator/good_client.c [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+ * Example use of bitmap allocator, showing how
+ * 'good' client code should interact with the
+ * allocator.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "bitmap_alloc.h"
+// #include "include/common.h"
```
Comment (ltratt) [inline example_allocators/bitmap_allocator/good_client.c:16]: Commented code.

File: example_allocators/bitmap_allocator/good_client.c [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+{
+	unsigned int i = 0;
+	unsigned int n = 1;
+	list *l = NULL;
+
+	init_alloc(NUM_CELLS, sizeof(list));
+
+	while (n <= NUM_CELLS)
+	{
+
```
Comment (ltratt) [inline example_allocators/bitmap_allocator/good_client.c:41]: Extraneous blank line.

**Comment by ltratt** [comment]
I have one over-arching question which is: what does this allocator show over the other allocators in particular the bump allocator? Naively, this looks less capable, and less CHERIfied than the bump allocator?

File: example_allocators/bitmap_allocator/bitmap_alloc.c (hunk 2 above)
Comment (djichthys) [reply example_allocators/bitmap_allocator/bitmap_alloc.c:61]: This is set by clang-format. Should we change clang-format in this PR or should this be done separately ?

File: example_allocators/bitmap_allocator/bitmap_alloc.c (hunk 3 above)
Comment (djichthys) [reply example_allocators/bitmap_allocator/bitmap_alloc.c]: This is the older commit within the two commits. This has changed in the second commit.

File: example_allocators/bitmap_allocator/bitmap_alloc.c (hunk 4 above)
Comment (djichthys) [reply example_allocators/bitmap_allocator/bitmap_alloc.c]: This is the older commit within the two commits. This has changed in the second commit.

File: example_allocators/bitmap_allocator/good_client.c (hunk 5 above)
Comment (djichthys) [reply example_allocators/bitmap_allocator/good_client.c:16]: Done. Thanks

File: example_allocators/bitmap_allocator/good_client.c (hunk 6 above)
Comment (djichthys) [reply example_allocators/bitmap_allocator/good_client.c:41]: Done thanks!

Comment (ltratt) [reply example_allocators/bitmap_allocator/good_client.c:41]: Can you push your fixed commits (not force pushing!) and say "Fixed in XYZ" where "XYZ" is the git commit? Thanks!

Comment (djichthys) [reply example_allocators/bitmap_allocator/good_client.c]: Fixed in `7906aad`.

File: example_allocators/bitmap_allocator/good_client.c (hunk 5 above)
Comment (djichthys) [reply example_allocators/bitmap_allocator/good_client.c]: Fixed in `7906aad`.

**Comment by ltratt** [comment]
Please squash as you see fit.

File: example_allocators/bitmap_allocator/README.md (hunk 1 above)
Comment (djichthys) [reply example_allocators/bitmap_allocator/README.md]: Fixed in `7906aad`.

**Comment by djichthys** [comment]
Shall I squash and force push as well @ltratt ?

**Comment by ltratt** [comment]
Yes please. When we say "please squash", you may always assume it means "and then force push" :)

**Comment by djichthys** [comment]
Ready to be merged.

**Comment by ltratt** [comment]
Can we fix up the commit messages please?

1. Let's capitalise things properly.
2. The formatting of commit messages seems to have two blank lines at each point? Maybe a Windows newline problem?
3. The title of the second commit doesn't make sense (it's about more than formatting)

Please force push updates.

**Comment by djichthys** [comment]
Commit title and message changed. Please check and merge if satisfactory @ltratt

**Comment by ltratt** [comment]
Thanks!

bors r+

**Comment by djichthys** [comment]
Sorry @ltratt. Fixed the build errors on this PR and pushed again.  Please could you check if alright and merge again. Hopefully the build errors will not happen

**Comment by ltratt** [comment]
Please can you undo the force push (remember: only force push when asked), and add a new commit with your changes? I'll review those changes and, assuming they're OK, ask you to squash and I'll try bors'ing again.

**Comment by djichthys** [comment]
Ok. Undid the force push and reset it back to [`d818cf9`](https://github.com/capablevms/cheri-examples/commit/d818cf9c37ebcc431c3d651d7d684af683680257). 

Then pushed two commits  (`c44822b`) [https://github.com/capablevms/cheri-examples/pull/20/commits/c44822bd8c5a83afea39b0aba5a6f6f12ff7c593] and (`
a148ffa` )[https://github.com/capablevms/cheri-examples/pull/20/commits/a148ffad2830c1ea6fff5bcf13006e88180923db]. 

Sorry about the force push!

**Comment by djichthys** [comment]
Squashed and force-pushed now. Please merge @ltratt . Thanks

**Comment by djichthys** [comment]
In trying to fix the  buildbot error, I ran clang-format on the file **bitmap_alloc.c**.  The diff shows that the only difference is a re-ordering of the include files between : 
* cherintrin.h 
* cheri/cheri.h

To compile correctly, we need the above order. clang-format  re-orders them.  Either  we change the `.clang-format` options or fix the intrinsics header within the compiler ?

**Comment by ltratt** [comment]
Does that mean we have too old a version of the headers or clang-format? Or, alternatively, that our clang-format settings are too scatter-gun and we don't want them to run on anything except *our* code in *this* repo?

**Comment by jeremysinger** [comment]
in general, I think we would want headers to be included in alphabetical order - do you agree?
also - we decided we probably want either cheriintrin.h or cheric.h but not both ... should we refactor our code so we only use one of them?

**Comment by djichthys** [comment]
> Does that mean we have too old a version of the headers or clang-format? Or, alternatively, that our clang-format settings are too scatter-gun and we don't want them to run on anything except _our_ code in _this_ repo?

I am using the latest headers on master for RISC-V.  our clang-format settings are let to use LLVM style. This reordering is default for the LLVM style.  However, the work-around is to break them up into separate blocks.

**Comment by djichthys** [comment]
> in general, I think we would want headers to be included in alphabetical order - do you agree?
> also - we decided we probably want either cheriintrin.h or cheric.h but not both ... should we refactor our code so we only use one of them?

unfortunately, while I would prefer to use `cheric.h`, **cheri_representable_length()** is only available in *cheriintrin.h*

**Comment by ltratt** [comment]
@jeremysinger I *think* that what @djichthys is saying is that clang-format is formatting system headers outside our repository (i.e. which aren't under our control). Have I got that correct @djichthys? If so, we should probably adjust clang-format to not inspect those headers.

**Comment by djichthys** [comment]
> @jeremysinger I _think_ that what @djichthys is saying is that clang-format is formatting system headers outside our repository (i.e. which aren't under our control). Have I got that correct @djichthys? If so, we should probably adjust clang-format to not inspect those headers.

Sorry!  ... what I meant is that clang-format is just re-ordering the block of includes in *OUR* code, but this causes compilation errors.  The way around this is to re-order the header includes into two separate blocks  ie. 

```
//block-1 
#include <cheriintrin.h> 
//add a space below 

#include <cheri/cheric.h>
```

**Comment by ltratt** [comment]
> Sorry! ... what I meant is that clang-format is just re-ordering the block of includes in OUR code, but this causes compilation errors

Ah... ouch! OK, can you just push a commit that adds a blank line at the right point then?

**Comment by djichthys** [comment]
pushed for review ! "blech"

**Comment by ltratt** [comment]
If it works, it works :) Please squash.

**Comment by djichthys** [comment]
Squashed and pushed. Please merge @ltratt . Fingers crossed!

**Pull request merged**